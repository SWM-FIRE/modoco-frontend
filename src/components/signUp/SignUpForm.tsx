import styled from 'styled-components';
import media from 'src/styles/media';
import Avatar from './Avatar';
import Email from './Email';
import Nickname from './Nickname';
import Password from './Password';
import useSignUp from '../../hooks/useSignUp';

export default function SignUpForm() {
  const {
    inputs,
    onChange,
    onSubmit,
    onChangeAvatar,
    isValidEmail,
    isValidPassword,
    isDisable,
  } = useSignUp();
  const { avatar, nickname, email, password, passwordCheck } = inputs;
  return (
    <Container data-cy="register-container">
      <Title data-cy="register-title">회원가입</Title>
      <Form onSubmit={onSubmit} data-cy="register-form">
        <Avatar avatar={avatar} onChangeAvatar={onChangeAvatar} />
        <Nickname nickname={nickname} onChange={onChange} />
        <Email email={email} onChange={onChange} isValidEmail={isValidEmail} />
        <Password
          password={password}
          passwordCheck={passwordCheck}
          onChange={onChange}
          isValidPassword={isValidPassword}
        />
        <Submit disabled={isDisable()} data-cy="register-submit">
          회원가입
        </Submit>
      </Form>
    </Container>
  );
}

const Container = styled.div`
  background-color: #18181b;
  display: flex;
  flex-direction: column;
  width: 61rem;
  align-items: flex-start;
  margin: 0 auto;
  padding: 7.8rem 0 14.4rem 0;
  font-family: IBMPlexSansKRRegular;
  color: #f9fafb;
  font-size: 1.5rem;
  border-radius: 0.6rem;
  ${media.small} {
    width: 80%;
    padding: 3.6rem 0 4.2rem 0;
  }
`;

const Title = styled.h1`
  font-size: 4rem;
  color: #f9fafb;
  ${media.small} {
    font-size: 2.4rem;
  }
`;

const Form = styled.form`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  width: 100%;

  input:focus {
    outline: none;
  }
`;

const Submit = styled.button`
  height: 5.5rem;
  background-color: #f3f4f6;
  color: #111827;
  width: 100%;
  font-size: 1.5rem;
  font-family: IBMPlexSansKRRegular;
  cursor: pointer;
  margin-top: 5rem;
  border-radius: 1rem;
  :disabled {
    cursor: default;
    background-color: #a9afb8;
  }
  ${media.small} {
    margin-top: 3rem;
    height: 4.6rem;
    font-size: 1.2rem;
  }
`;
