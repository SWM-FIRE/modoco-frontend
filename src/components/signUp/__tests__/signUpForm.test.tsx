import React from 'react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter as Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import SignUpForm from '../SignUpForm';

function renderSignUp() {
  // mocking submit button -> not used
  const onSubmit = jest.fn();

  // render SignUpForm
  const result = render(
    <Router>
      <SignUpForm />
    </Router>,
  );

  // get blocks
  const Heading = () => result.getByText('회원가입', { selector: 'h1' });
  const Nickname = () => result.getByPlaceholderText('닉네임을 입력해주세요.');
  const Email = () => result.getByPlaceholderText('이메일을 입력해주세요.');
  const FirstPassword = () =>
    result.getByPlaceholderText(
      '8~16자 영문 대 소문자, 숫자, 특수문자를 사용하세요.',
    );
  const SecondPassword = () =>
    result.getByPlaceholderText('비밀번호를 다시 입력해주세요.');

  // make events for blocks
  const typeNickname = (name: string) => {
    userEvent.type(Nickname(), name);
  };
  const typeEmail = (name: string) => {
    userEvent.type(Email(), name);
  };
  const typeFirstPassword = (pass: string) => {
    userEvent.type(FirstPassword(), pass);
  };
  const typeSecondPassword = (pass: string) => {
    userEvent.type(SecondPassword(), pass);
  };
  const SignUpButton = () =>
    result.getByText('회원가입', { selector: 'button' });
  const clickSubmit = () => {
    userEvent.click(SignUpButton());
  };

  return {
    result,
    Heading,
    Nickname,
    Email,
    FirstPassword,
    SecondPassword,
    typeNickname,
    typeEmail,
    typeFirstPassword,
    typeSecondPassword,
    SignUpButton,
    clickSubmit,
    onSubmit,
  };
}

describe('<SignUpForm />', () => {
  it('sign up form rendering', () => {
    const {
      Heading,
      Nickname,
      Email,
      FirstPassword,
      SecondPassword,
      SignUpButton,
    } = renderSignUp();

    // check if blocks are rendered
    expect(Email()).toBeInTheDocument();
    expect(Nickname()).toBeInTheDocument();
    expect(Heading()).toBeInTheDocument();
    expect(FirstPassword()).toBeInTheDocument();
    expect(SecondPassword()).toBeInTheDocument();
    expect(SignUpButton()).toBeInTheDocument();
  });

  it('form 값으로 onSubmit을 호출.', async () => {
    const {
      Email,
      Nickname,
      FirstPassword,
      SecondPassword,
      typeNickname,
      typeEmail,
      typeFirstPassword,
      typeSecondPassword,
      clickSubmit,
    } = renderSignUp();

    // make events for blocks
    typeNickname('영기');
    typeEmail('yeonggi@mail.com');
    typeFirstPassword('somatest0909!');
    typeSecondPassword('somatest0909!');
    clickSubmit();

    // check if events are working
    expect(Email()).toContainHTML('yeonggi@mail.com');
    expect(Nickname()).toContainHTML('영기');
    expect(FirstPassword()).toContainHTML('somatest0909!');
    expect(SecondPassword()).toContainHTML('somatest0909!');
  });
});
