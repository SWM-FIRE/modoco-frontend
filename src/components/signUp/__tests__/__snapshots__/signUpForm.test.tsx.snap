// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUpForm /> matches snapshot 1`] = `
<div>
  <div
    class="sc-ksZaOG bqwFvM"
  >
    <h1
      class="sc-hAZoDl eIvRUQ"
    >
      회원가입
    </h1>
    <form
      class="sc-fnykZs iFTbmL"
    >
      <div
        class="sc-bczRLJ fqGdDf"
      >
        <label
          class="sc-dkzDqf Revln"
        >
          프로필 이미지
        </label>
        <div
          class="sc-gsnTZi ddTkGA"
        >
          <svg
            fill="none"
            height="80"
            role="img"
            viewBox="0 0 36 36"
            width="80"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Mahalia Jackson
            </title>
            <mask
              height="36"
              id="mask__beam"
              maskUnits="userSpaceOnUse"
              width="36"
              x="0"
              y="0"
            >
              <rect
                fill="#FFFFFF"
                height="36"
                rx="72"
                width="36"
              />
            </mask>
            <g
              mask="url(#mask__beam)"
            >
              <rect
                fill="#ffad08"
                height="36"
                width="36"
              />
              <rect
                fill="#73b06f"
                height="36"
                rx="36"
                transform="translate(2 6) rotate(222 18 18) scale(1)"
                width="36"
                x="0"
                y="0"
              />
              <g
                transform="translate(-6 1) rotate(-2 18 18)"
              >
                <path
                  d="M13,19 a1,0.75 0 0,0 10,0"
                  fill="#000000"
                />
                <rect
                  fill="#000000"
                  height="2"
                  rx="1"
                  stroke="none"
                  width="1.5"
                  x="12"
                  y="14"
                />
                <rect
                  fill="#000000"
                  height="2"
                  rx="1"
                  stroke="none"
                  width="1.5"
                  x="22"
                  y="14"
                />
              </g>
            </g>
          </svg>
          <button
            class="sc-hKMtZM fPBanq"
            type="button"
          >
            아바타 재생성
          </button>
        </div>
      </div>
      <p
        class="sc-ftvSup jzktyu"
      >
        <label
          class="sc-papXJ hWwCsK"
          for="nickname"
        >
          닉네임 *
        </label>
        <input
          class="sc-jqUVSM fmIAbp"
          id="nickname"
          name="nickname"
          placeholder="닉네임을 입력해주세요."
          type="text"
          value=""
        />
        <span
          class="sc-kDDrLX iNbjoI"
        />
      </p>
      <p
        class="sc-eCYdqJ fmuhWq"
      >
        <label
          class="sc-jSMfEi cbffqr"
          for="email"
        >
          이메일 *
        </label>
        <input
          class="sc-gKXOVf iMMMEw"
          id="email"
          name="email"
          placeholder="이메일을 입력해주세요."
          type="text"
          value=""
        />
        <span
          class="sc-iBkjds cgqYHT"
        />
      </p>
      <p
        class="sc-iqcoie hKPtSU"
      >
        <label
          class="sc-crXcEl jyaNby"
          for="password"
        >
          비밀번호 *
        </label>
        <input
          class="sc-evZas jZjSzQ"
          id="password"
          name="password"
          placeholder="8~16자 영문 대 소문자, 숫자, 특수문자를 사용하세요."
          type="password"
          value=""
        />
        <span
          class="sc-breuTD ersDwp"
        />
      </p>
      <p
        class="sc-iqcoie hKPtSU"
      >
        <label
          class="sc-crXcEl jyaNby"
          for="passwordCheck"
        >
          비밀번호 확인 *
        </label>
        <input
          class="sc-evZas jZjSzQ"
          id="passwordCheck"
          name="passwordCheck"
          placeholder="비밀번호를 다시 입력해주세요."
          type="password"
          value=""
        />
        <span
          class="sc-breuTD ersDwp"
        />
      </p>
      <button
        class="sc-fEOsli gsYZga"
        disabled=""
      >
        회원가입
      </button>
    </form>
  </div>
</div>
`;
